File: returned/python/longest_substring_without_repeating_characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) <= 1:
            return len(s)
        maximum = 1
        l = 0
        r = 1
        cache = set(s[0])
        while r <= len(s) - 1:
            while (s[r] in cache) & (l < r):
                cache.remove(s[l])
                l += 1
            cache.add(s[r])
            maximum = max(maximum, r - l + 1)
            r += 1
        return maximum

File: typescript/solved/longest-substring-without-repeating-characters.ts
/**
 * https://leetcode.com/problems/longest-substring-without-repeating-characters/description/
 *
 * Given a string s, find the length of the longest substring without repeating characters.
 *
 * Solution: [NeetCode - Longest Substring Without Repeating Characters - Leetcode 3 - Python](https://www.youtube.com/watch?v=wiGpQwVHdE0)
 */

export function lengthOfLongestSubstring(s: string): number {
  const set = new Set();
  let index = 0;
  let max = 0;
  for (let i = 0; i < s.length; i++) {
    const char = s[i];
    while (set.has(char)) {
      set.delete(s[index]);
      index++;
    }
    set.add(char);
    max = Math.max(set.size, max);
  }
  return max;
}
